src/mito2/src/worker.rs

```diff
@@ -581,6 +592,59 @@ impl<S: LogStore> RegionWorkerLoop<S> {
             let _ = sender.send(SetReadonlyResponse::NotFound);
         }
     }
+
+    /// Handles `set_writable_gracefully`.
+    async fn set_writable_gracefully(
+        &mut self,
+        region_id: RegionId,
+        expected_last_entry_id: Option<EntryId>,
+        sender: oneshot::Sender<SetWritableResponse>,
+    ) {
+        if let Some(region) = self.regions.get_region(region_id) {
+            if region.is_writable() {
+                let _ = sender.send(SetWritableResponse::Result(Ok(())));
+                return;
+            }
+
+            let flushed_entry_id = region.version_control.current().last_entry_id;
+
+            // TODO(weny): What if the region manifest is not the latest version?
+            // Consider adding tests for manifest replaying.
+            match replay_memtable(
+                &self.wal,
+                region_id,
+                flushed_entry_id,
+                &region.version_control,
+            )
+            .await
```

Comment:
Simply replaying the memtable might lead to an incorrect state. The writable region might flush and bump wal entries after `last_entry_id` before a new node sets the region to writable. We should always reopen the region.